### PR TITLE
Answer in Docker in GPU with all the fields

### PR DESCRIPTION
**Pull request by bryant1410**
I added support to run Docker in GPU and added the missing fields.

I'll leave more comments in the code.

Note Flask in debug mode shows the output twice because it forks itself to be ready for reloading. See https://stackoverflow.com/a/25504196/1165181

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -36,8 +36,12 @@ jobs:
...
+      run: |
+        source activate covid19-qa
+        python corpus-creator/corpus_creator.py
+    - name: Run QA main
       run: |
-        cd qa
         source activate covid19-qa
-        python main.py --ignore-es=True
+        cd qa
+        python main.py --ignore-es
```
Comment (bryant1410) [inline .github/workflows/main.yml:47]: Change the flag to be an "action_store"

File: docker-compose.override.yml [hunk 2]
```diff
@@ -0,0 +1,45 @@
+version: "3.8"
```
Comment (bryant1410) [inline docker-compose.override.yml:1]: This file is for debugging purposes, which is loaded by default by Docker Compose.

File: docker-compose.prod.yml [hunk 3]
```diff
@@ -0,0 +1,18 @@
+version: "3.8"
```
Comment (bryant1410) [inline docker-compose.prod.yml:1]: This file is an override for production.

File: docker-compose.yml [hunk 4]
```diff
@@ -1,84 +1,82 @@
-version: "3.7"
+version: "3.8"
```
Comment (bryant1410) [inline docker-compose.yml:1]: I only left in this file things that should be set up regardless of the env.

File: kibana/config/kibana.yml [hunk 5]
```diff
@@ -3,5 +3,5 @@
 #
 server.name: kibana
 server.host: "0"
-elasticsearch.hosts: [ "http://elasticsearch-covid:9200" ]
-xpack.monitoring.ui.container.elasticsearch.enabled: true
\ No newline at end of file
+elasticsearch.hosts: [ "http://elasticsearch:9200" ]
```
Comment (bryant1410) [inline kibana/config/kibana.yml:6]: I removed all the "-covid" suffixes because they're noisy and redundant IMHO.

File: load_test.sh [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+set -ex
+
+url=http://localhost:8000
+
+for i in {1..10}; do
+  # Different questions so they are cache-missed.
+  curl -sS --data "question=¿Qué dijo $i?" -X POST "${url}/question/" -H "accept: application/json" > /dev/null &
+done
+
+wait
```
Comment (bryant1410) [inline load_test.sh:12]: This file sends 10 requests at the same time. It's kinda a load test.

File: qa/Dockerfile [hunk 7]
```diff
@@ -1,7 +1,23 @@
...
+# (largest batch size possible; we try to check with all the document fragments we can).
+# We then can just run this sync.
+#
+# So, it makes sense just to use Flask server in production.
+# The problems with flask in production are about the response time and scalability,
+# which we don't care because we're just running sync.
+# Another problem could be the security risk of enabling debug mode in production.
+# Let's just avoid that with the env var and that's it.
+# See https://stackoverflow.com/a/12269934/1165181
+CMD source activate covid19-qa && flask run -h 0.0.0.0
```
Comment (bryant1410) [inline qa/Dockerfile:23]: A disadvantage I found from doing `conda run` is that the stderr is never shown (or heavily buffered to the point that I never see it...). This way, we can see the stderr.

File: qa/Dockerfile [hunk 8]
```diff
@@ -1,7 +1,23 @@
-FROM continuumio/miniconda3
-WORKDIR /src
+FROM pytorch/pytorch:1.5-cuda10.1-cudnn7-runtime
```
Comment (bryant1410) [inline qa/Dockerfile:1]: This image contains conda, pytorch, cuda, cudnn

File: qa/Dockerfile (hunk 7 above)
Comment (bryant1410) [reply qa/Dockerfile:23]: See this explanation above on why to use Flask for production here. I found hard to use others.

File: qa/api/views.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+            return [
+                {
+                    "title": answer.instance.document.title,
+                    "context": answer.instance.context_text,
+                    "answer": answer.text,
+                    "answer_start_index": answer.start,
+                    "answer_end_index": answer.end,
+                    "date": answer.instance.document.date,
+                    "source": answer.instance.document.source,
+                    "url": answer.instance.document.url,
```
Comment (bryant1410) [inline qa/api/views.py:48]: All the fields are tied now.

File: qa/covid19_qa/elasticsearch_qa.py [hunk 10]
```diff
@@ -1,49 +1,47 @@
...
+from typing import Any, Iterator, Mapping
+
 from elasticsearch import Elasticsearch
-from typing import Iterator
 
-from covid19_qa.pipeline import Instance
+from covid19_qa.pipeline import Instance, Document
+
+ES_HOST = os.getenv("ES_HOST")
+es = Elasticsearch([{"host": ES_HOST, "port": os.environ["ES_PORT"]}]) if ES_HOST else None
```
Comment (bryant1410) [inline qa/covid19_qa/elasticsearch_qa.py:9]: It's parametrized now.

File: qa/covid19_qa/pipeline.py [hunk 11]
```diff
@@ -16,7 +16,23 @@
...
+    source: str
+    url: str
+
+
+class Instance(SquadExample):
+    # Sometimes we're just asking questions without testing, mostly when we evaluate.
+    # So we can set the document to null then.
+    def __init__(self, document: Optional[Document] = None, **kwargs):
+        super().__init__(title=document.title, **kwargs)
+        self.document = document
```
Comment (bryant1410) [inline qa/covid19_qa/pipeline.py:35]: In the end, subclassing worked fine. I feared some Transformers library methods but proved pointless. We keep a document here, which now has a title, URL, source, and date.

File: qa/covid19_qa/pipeline.py [hunk 12]
```diff
@@ -96,7 +112,7 @@ def __call__(self, *texts, **kwargs) -> Iterator[Answer]:
         kwargs.setdefault("max_answer_len", 15)
         kwargs.setdefault("max_seq_len", 384)
         kwargs.setdefault("max_question_len", 64)
-        kwargs.setdefault("version_2_with_negative", False)
+        kwargs.setdefault("handle_impossible_answer", False)
```
Comment (bryant1410) [inline qa/covid19_qa/pipeline.py:115]: Just keep up-to-date with some naming changed upstream.

File: qa/covid19_qa/qa.py [hunk 13]
```diff
@@ -36,16 +37,16 @@ def answer_from_instances(instances: Iterator[Instance], qa_pipeline: Pipeline,
     else:
         return heapq.nlargest(top_k, answers, key=lambda a: a.sort_score)
 
+
+@lru_cache
```
Comment (bryant1410) [inline qa/covid19_qa/qa.py:41]: We cache up to 128 input parameter configurations (the default), which should be only diff questions. We can change this number later, or even make it an env var.

Comment (bryant1410) [reply qa/covid19_qa/qa.py:41]: You can test this by making the same request twice, and noting is way faster after the first time.

File: README.md [hunk 14]
```diff
@@ -38,65 +37,125 @@ python manage.py createsuperuser
...
+    docker-compose up kibana
+    ```
+
+2. In your browser, go to http://0.0.0.0:15601/app/kibana
+
+## Make Docker work with Nvidia
+
+The `qa` image is CUDA-enabled. It needs to run with the `nvidia` runtime to work well.
+
+1. Install [nvidia-docker-runtime](https://github.com/NVIDIA/nvidia-container-runtime).
```
Comment (bryant1410) [inline README.md:117]: For just Docker, this is enough. But not for Docker Compose, unfortunately. For it, we need the next step below.

**Comment by raulsperoni** [comment]
Las cosas que entiendo las veo bien

**Comment by pgrill** [comment]
Great work Santiago 👏 👏

**Pull request merged**